common/log.h

```diff
@@ -456,3 +481,86 @@ inline void log_print_usage()
     fprintf(stdout, "  --log-file            Specify a log filename (without extension)\n");
     fprintf(stdout, "                        Log file will be tagged with unique ID and written as \"<name>.<ID>.log\"\n"); /*  */
 }
+
+inline void log_dump_cmdline(int argc, char **argv)
+{
+    std::string buf;
+    for (int i = 0; i < argc; ++i)
+    {
+        if (std::string(argv[i]).find(' ') != std::string::npos)
+        {
+            buf.append(" \"").append(argv[i]).append("\"");
+        }
+        else
+        {
+            buf.append(" ").append(argv[i]);
+        }
+    }
+    LOGLN("Cmd:%s", buf.c_str())
+}
+
+#define LOG_TOSTR(var) _log_var_to_string(var).c_str()
+
+inline std::string _log_var_to_string(bool var)
+{
+    return var ? "true" : "false";
+}
+
+inline std::string _log_var_to_string(std::string var)
+{
+    return var;
+}
+
+inline std::string _log_var_to_string(std::vector<int> var)
+{
+    std::string buf;
+    buf.append("[ ");
+    bool first = true;
+    for (auto e : var)
+    {
+        if (first)
+        {
+            first = false;
+        }
+        else
+        {
+            buf.append(", ");
+        }
+        buf.append(std::to_string(e));
+    }
+    buf.append(" ]");
+
+    return buf;
+}
+
+#define LOG_TOKENS_TOSTR_PRETTY(tokens, ctx)                        \
+    [&tokens, &ctx]()                                               \
+    {                                                               \
+        std::string buf("[ ");                                      \
+        bool first = true;                                          \
+        for (const auto &token : tokens)                            \
+        {                                                           \
+            if (!first)                                             \
+                buf.append(", ");                                   \
+            else                                                    \
+                first = false;                                      \
+                                                                    \
+            auto detokenized = llama_token_to_str(ctx, token);      \
+                                                                    \
+            detokenized.erase(                                      \
+                std::remove_if(                                     \
+                    detokenized.begin(),                            \
+                    detokenized.end(),                              \
+                    [](const char c) { return !std::isprint(c); }), \
+                detokenized.end());                                 \
+                                                                    \
+            buf                                                     \
+                .append("'")                                        \
```

Comment:
I think this kind of use case is exactly what std::stringstream was made for.

Reply:
I had a really exhausting day at work today, so I just pushed what I had before my two remaining brain cells start reinventing the wheel :)

But suggestion much appreciated.